BLD: Intel distutils fixes

dzagorny opened the pull request:
Fixes:
- Error `failed with exit status -1073741502` during build Scipy
- MSVCCompiler overwrite 'lib' and 'include' environment variables (including Intel environment)
- Issue with Scipy build in case, if python installed into folder with whitespaces

charris commented on numpy/distutils/fcompiler/compaq.py:
```diff
@@ -74,7 +74,7 @@ class CompaqVisualFCompiler(FCompiler):
 
     if sys.platform=='win32':
-        from distutils.msvccompiler import MSVCCompiler
+        from numpy.distutils.msvccompiler import MSVCCompiler
```
This line was due to Intel just a few days ago, that make me suspicious of this fix.

charris replied:
And I see you are Intel also. Hmm...

dzagorny replied:
I know about Yolanda's pull request, it is addition part of fixes.

charris commented:
This is far from a minimal patch and numpy distutils is pretty fragile, so it makes me nervous. Also, see `doc/source/dev/gitwash/development_workflow.rst` for guidelines about writing commit messages. Also keep all lines, including the summary line, <= 72 characters in length.

dzagorny commented:
- Added closing for 'so_dup' and 'se_dup' file descriptors - you can see in code, that ```so_dup``` and ```se_dup``` not used after closed.
- MSVCCompiler overwrite 'lib' and 'include'... - this patch affect Intel Compilers and Compaq Compiler. Changes are minimal, now environment added, not overwritten:
```os.environ['lib'] = environ_lib + os.environ['lib']```
```os.environ['include'] = environ_include + os.environ['include']```
- Fixed issue: SciPy can't be build... - May affect all compilers in case, if compilation flags are incorrect(contain unexpected whitespaces)

antalmal commented on numpy/distutils/intelccompiler.py:
```diff
@@ -15,14 +17,15 @@ class IntelCCompiler(UnixCCompiler):
-                             linker_exe=compiler,
-                             linker_so=compiler + ' -shared')
+                             linker_exe=compiler + ' -shared-intel',
+                             linker_so=compiler + ' -shared -shared-intel')
```
Are you sure? it will lead to dependency on shared libraries from Intel so that resulting application become less portable.

dzagorny replied:
I'm not sure, that statically linking with compiler runtime is a good variant. But in another hand shared libraries...

charris replied:
@cgohlke Does this look correct to you?

charris commented:
@cgohlke, @cournape, @ilanschnell Any chance you folks could take a look at this or test it? I'd like it to be in 1.10 but it makes me a bit nervous.

charris commented on numpy/distutils/msvc9compiler.py:
```diff
@@ -0,0 +1,22 @@
+import os
+from distutils.msvccompiler import *
```
It is best to avoid `import *`, instead either use the namespace, `distutils.msvccompiler.whatever`, which has the advantage of making it clear where everything comes from, or explicitly import those objects needed.

dzagorny replied:
@charris msvc9compiler.py serves as decorator for several methods in ```distutils.msvc9compiler.MSVCCompiler```. But distutils.msvc9compiler.py contains not only ```MSVCCompiler``` class. ```<python>\...\distutils\msvc9compiler.py``` also contains another classes, functions, variables. ```import *``` is allowed to use these objects from ```<numpy>\distutils\msvc9compiler.py```.

charris commented on numpy/distutils/msvccompiler.py:
```diff
@@ -0,0 +1,17 @@
+import os
+from distutils.msvccompiler import *
```
See above about `*` imports. I think it would be clearer to use `distutils.MSVCCompiler`, why mess with an `_`.

charris commented:
The commit messages should have hard line breaks and lines should be <=72 characters in length. There are no line breaks in the current commit messages. Also see `doc/source/dev/gitwash/development_workflow.rst` for proper summary line formatting. You can rewrite/squash you commits using `git rebase -i`.

charris commented:
@dzagorny I'd like to have this in 1.10, but it needs some work.

dmitrii-zagornyi committed:
Added closing for 'so_dup' and 'se_dup' file descriptors.

Issue: _exec_command function doesn't close 'so_dup' and 'se_dup' file
descriptors.SciPy try to build scipy\special\amos\zunik.f and
crash:error: Command "C:\Program Files (x86)\Intel\Composer XE
2015\bin\intel64\ifort.exe /nologo /MD /nbs /names:lowercase
/assume:underscore /O1
-IC:\Python27\lib\site-packages\numpy\core\include -c /c
scipy\special\amos\zunik.f
/Fobuild\temp.win-amd64-2.7\scipy\special\amos\zunik.o" failed with
exit status -1073741502

dmitrii-zagornyi committed:
MSVCCompiler overwrite 'lib' and 'include' environment variables. This
behavior affect at least python 3.5 and SciPy build and build failed.
During initialization <python>.distutils.MSVCCompiler replace Intel
environment('include' and 'lib' paths). This fix decorate 'initialize'
function in MSVCCompiler and extend 'lib' and 'include' environment
variables. Changed compilation keys: generate optimized code
specialized for Intel processors with SSE4.2 support.

dmitrii-zagornyi committed:
Fixed issue: SciPy can't be build in case, if python was installed into
folder with whitespaces.

dmitrii-zagornyi committed:
Align with pep8

dmitrii-zagornyi committed:
Changed from distutils_MSVCCompiler to
distutils.msvccompiler.MSVCCompiler

dzagorny commented:
Hm, I added line breaks, but looks like comments still haven't line breaks. Could you check?

charris commented:
Still no line breaks. How are you writing the commit messages?

dzagorny commented:
I use tortoise git. I think, now commit messages should be <=72 characters in length.

charris commented:
Still a problem. You are working on windows, correct? I suspect a line ending problem but can't see why it would have this affect. It would be weird if tortoisegit didn't have insert line breaks at all.

dzagorny commented:
I'm working on windows. But now I checked on linux and I see, that my commits have line breaks:
```
commit 5ef26af9209f206c36329e0b2585b4ebdbb5f5fd
Author: Dmitry Zagorny
Date:   Mon Sep 14 13:07:17 2015 -0500

    Changed from distutils_MSVCCompiler to
    distutils.msvccompiler.MSVCCompiler
```
Are you sure, that you look into last version?

charris commented:
I pull down the commits as patches and apply them (git am). The first patch seems to have some linebreaks, albeit in strange places, but the remainder are missing.

![screenshot from 2015-09-15 10-48-26](https://cloud.githubusercontent.com/assets/77272/9883537/e2ec6410-5b97-11e5-937b-e2631b49a6f9.png)

dzagorny commented:
Could you update your local branch?
Your screenshot contain old commits. For example last commit ```Changed from distutils_MSVCCompiler to``` should have this hash:
5ef26af9209f206c36329e0b2585b4ebdbb5f5fd

charris commented:
The hashes will not be the same as I'm not pulling your branch, but pulling down patches with `curl https://patch-diff.githubusercontent.com/raw/numpy/numpy/pull/6243.patch`. I'll check out your branch and see if anything changes.

charris commented:
OK, if I check out your branch the line breaks are there. I don't know what is going on with the patches but suspect a setting someplace. Never had this problem with other PRs.

dzagorny commented:
Do I understand right, that now comments are ok?

charris commented:
Yes, the comments look OK when I pull your branch directly. May have to merge to see what the final result is :-(

charris commented:
OK, lets give this a shot. @dzagorny For next time, take a look at `doc/source/dev/gitwash/development_workflow.rst` for instruction on formatting commit messages.
